### PR TITLE
Speed up prepare_roads()

### PR DESCRIPTION
**Pull request by sebastiantimwagner**
### Problem Statement

1. In the current version, `prepare.py` uses a custom function in `_util` called `clip_line_poly()` which was discovered to be rather slow. It also turned out to be a bottleneck in the execution of `example.ipynb`. In this instance, the function is used to clip the roads dataset by an Area of Interest, which is the country of Burundi in `example.ipynb`

2. The example road dataset does not contain a `power` column, which is currently not checked in the function during preprocessing.

### Solution

1. Use built-in functions in geopandas that read in a vector dataset with a mask. Since this masked file reading can result in some road geometries that lie outside of the boundary of the mask, a subsequent spatial join needs to be executed to clip the data correctly. This approach is significantly faster than the previous approach. 

2. Add check whether `power` column exists and if not the case, continue.

### CODE REVIEW

File: gridfinder/prepare.py [hunk 1]
```diff
@@ -336,8 +336,9 @@ def prepare_roads(roads_in, aoi_in, ntl_in, include_power=True):
     else:
         aoi = gpd.read_file(aoi_in)
 
-    roads = gpd.read_file(roads_in)
-    roads = clip_line_poly(roads, aoi)
+    roads_masked = gpd.read_file(roads_in, mask=aoi)
+    roads = gpd.sjoin(roads_masked, aoi, how="inner", op="intersects")
```
Comment (carderne) [inline gridfinder/prepare.py:340]: ~This shouldn't be necessary to get equivalent behaviour. `roads` is immediately rasterized so any extra columns you get are being thrown away. Am I missing something?~

Doh, just read what you said about the lines lying outside the AOI.

File: gridfinder/prepare.py [hunk 2]
```diff
@@ -350,7 +351,8 @@ def prepare_roads(roads_in, aoi_in, ntl_in, include_power=True):
     roads.loc[roads["highway"] == "service", "weight"] = 1 / 3
 
     # Power lines get weight 0
-    roads.loc[roads["power"] == "line", "weight"] = 0
+    if "power" in roads:
```
Comment (carderne) [inline gridfinder/prepare.py:354]: Great

**Review by carderne — CHANGES_REQUESTED** [review]
Should make the same change at https://github.com/carderne/gridfinder/blob/d51b269fcf1bc28bef0880bf432e97931411981d/gridfinder/post.py#L199 and remove the `clip_line_poly()` function altogether.

**Comment by sebastiantimwagner** [comment]
> Should make the same change at
> 
> https://github.com/carderne/gridfinder/blob/d51b269fcf1bc28bef0880bf432e97931411981d/gridfinder/post.py#L199
> and remove the `clip_line_poly()` function altogether.

Requested changes now resolved in https://github.com/carderne/gridfinder/pull/5/commits/610eb6796a9e75c79c43e4b8d94015c1c28e1c15

File: gridfinder/post.py [hunk 3]
```diff
@@ -195,8 +193,10 @@ def accuracy(grid_in, guess_in, aoi_in, buffer_amount=0.01):
     else:
         aoi = gpd.read_file(aoi_in)
 
-    grid = gpd.read_file(grid_in)
-    grid_clipped = clip_line_poly(grid, aoi)
+    grid_masked = gpd.read_file(grid_in, mask=aoi)
+    grid_clipped = gpd.sjoin(grid_masked, aoi, how="inner", op="intersects")
+    grid_clipped = grid_clipped[grid_clipped.columns]
```
Comment (carderne) [inline gridfinder/post.py]: This should follow the same pattern as in `prep_roads()`. Also as it is this won't drop any columns.
Should be:
```
grid_masked = gpd.read_file(grid_in, mask=aoi)
grid = gpd.sjoin(grid_masked, aoi, how="inner", op="intersects")
grid = grid[grid_masked.columns]
```

**Comment by carderne** [comment]
Just added one quick comment will merge once that's done.

And a request: until we figure out how you guys are going to take more control of development, could you commit your changes in a separate branch, so we don't get previous commits and merges tagging along in the PR. :smile:

**Pull request merged**